File: matlab/logmart.m
%solve y=Ax using parallel log-entropy MART  (De Pierro 1991)
function [ x,y_est,chi2,i ] = logmart( y,A,relax,x0,sigma,max_iter )
% Program is stopped if Chisquare increases.
% A is NxM array
% Y is Nx1 vector
% returns Mx1 vector
%
% KEYWORDS:
% relax		user specified relaxation constant	(default is 20.)
% x0	user specified initial guess (N vector)
%           (default is backproject y, i.e., y#A)
% max_iter	user specified max number of iterations (default is 20)
%
%AUTHOR:	Joshua Semeter
%LAST MODIFIED:	5-2015

%  Simple test problem
% A=diag([5 5 5]);
% x=[1;2;3];
% y=A*x;
narginchk(2,6)
assert(iscolumn(y),'y must be a column vector')
validateattributes(y, {'numeric'}, {'nonnegative', 'vector'})
validateattributes(A, {'numeric'}, {'nonnegative', 'ndims', 2})
assert(size(A,1)==size(y,1),'A and y row numbers must match')
%% set defaults
if (nargin<6), max_iter=200.; end
if (nargin<5), sigma=1.; end
%% make sure there are no 0's in y
y(y<=1e-8)=1e-8;

if (nargin<4) || isempty(x0)
    x=(A'*y)./sum(A(:));
    xA=A*x;
    x=x.*max(y(:))/max(xA(:));
%    max(x(:))

else
    x=x0;
end
if (nargin<3), relax=1; end
validateattributes(x0, {'numeric'}, {'nonnegative'})
validateattributes(relax, {'numeric'}, {'scalar', 'positive'})
validateattributes(sigma, {'numeric'}, {'scalar', 'positive'})
validateattributes(max_iter, {'numeric'}, {'scalar', 'positive'})


% W=sigma;
% W=linspace(1,0,size(A,1))';
% W=rand(size(A,1),1);
W=ones(size(A,1),1);
W=W./sum(W);

chi2 = chi_squared(y, A, x, sigma);

for i = 1:max_iter
%%  iterate solution, plot estimated data (diag elems of x#A)
  xold=x;
  t=min(1./(A*x));
  C=(relax*t*( 1-((A*x)./y) ));
  x=x./(1-x.*(A'*(W.*C)));
%% monitor solution
  chiold=chi2;
  chi2 = chi_squared(y, A, x, sigma);
%   dchi2=(chi2-chiold);
  if ((chi2>chiold) && (i>2)), break, end  %  || (chi2<0.7)

%figure(9); clf; hold off;
%Nest=reshape(x,69,83);
%imagesc(Nest); caxis([0,1e11]);
%set(gca,'YDir','normal'); set(gca,'XDir','normal');
%pause;

end % for
x=xold;
if nargout>1
    y_est=(A*x);
end
end %function

function chi2 = chi_squared(y, A, x, sigma)
chi2 = sqrt( sum(((A*x - y)./sigma).^2) );
end

File: fortran/logmart.f90
module art

use iso_fortran_env, only: wp=>real64

implicit none

contains

pure subroutine logmart(A,b,relax,x0,sigma,max_iter, x)
! delta Chisquare.
! stopped if Chisquare increases.
!
! Inputs
! ------
! A: NxM array
! b: N vector
! relax: user specified relaxation constant	(default is 20.)
! x0: user specified initial guess (Nx1 vector)  (default is backproject y, i.e., y#A)
! max_iter: user specified max number of iterations (default is 20)
!
!
! Outputs
! -------
! x: N vector
!
!
!    AUTHOR:	Joshua Semeter
!    LAST MODIFIED:	5-2015
!
!      Simple test problem
!    A = diag([5, 5, 5])
!    x = [1,2,3]
!    b = A*x

! --- parameter check
real(wp), intent(in) :: A(:,:), b(:)
real(wp), optional, value :: relax, sigma
real(wp), intent(in),optional :: x0(:)
integer, optional, value :: max_iter
real(wp), intent(out) :: x(:)

real(wp), dimension(size(b)) :: W(size(b)), x_prev,c, op_b
integer :: i
real(wp) :: t,chi2,chiold


if (.not.size(A,1) == size(b)) error stop 'A and b row numbers must match'
if (any(A<0)) error stop 'A must be non-negative'
if (any(b<0)) error stop 'b must be non-negative'
op_b = b
! --- make sure there are no 0's in b
where(op_b <= 1e-8) op_b = 1e-8_wp

! --- set defaults
if (.not.present(relax)) relax = 1
if (.not.present(max_iter)) max_iter = 200
if (.not.present(sigma))  sigma = 1

if (.not.present(x0)) then
  x  = matmul(transpose(A), op_b) / sum(A)
  x  = x * maxval(op_b) / maxval(matmul(A, x))
else
  x = x0
endif



! W=sigma;
! W=linspace(1,0,size(A,1))';
! W=rand(size(A,1),1);
W = 1
W = W / sum(W)

! --- iterate solution
chi2 = chi_squared(A, op_b, x, sigma)

do i = 1, max_iter
  x_prev = x
  t = minval(1/matmul(A,x))
  C = relax*t*(1-(matmul(A,x)/op_b))
  x = x / (1-x*matmul(transpose(A),W*C))
! monitor solution
  chiold = chi2
  chi2 = chi_squared(A, op_b, x, sigma)
  if (chi2 > chiold .and. i > 2) exit
enddo

x = x_prev

end subroutine logmart


pure real(wp) function chi_squared(A, b, x, sigma)
real(wp), intent(in) :: A(:,:), b(:), x(:), sigma
chi_squared = sqrt(sum(((matmul(A,x) - b) / sigma)**2))

end function chi_squared

end module art
